pkg/distributor/query.go

```diff
@@ -232,22 +255,53 @@ func (d *Distributor) queryIngesterStream(ctx context.Context, replicationSet ri
 					}
 				}
 
-				if chunkBytesLimitErr := queryLimiter.AddChunkBytes(resp.ChunksSize()); chunkBytesLimitErr != nil {
+				if chunkBytesLimitErr := queryLimiter.AddChunkBytes(ingester_client.ChunksSize(resp.Chunkseries)); chunkBytesLimitErr != nil {
 					return ingesterQueryResult{}, validation.LimitError(chunkBytesLimitErr.Error())
 				}
 
 				result.chunkseriesBatches = append(result.chunkseriesBatches, resp.Chunkseries)
+			} else if len(resp.Series) > 0 {
+				labelsBatch := make([]labels.Labels, 0, len(resp.Series))
+				streamingSeriesCount += len(resp.Series)
+
+				for _, s := range resp.Series {
+					if limitErr := queryLimiter.AddSeries(s.Labels); limitErr != nil {
+						return ingesterQueryResult{}, validation.LimitError(limitErr.Error())
+					}
+
+					labelsBatch = append(labelsBatch, mimirpb.FromLabelAdaptersToLabels(s.Labels))
+				}
+
+				streamingSeriesBatches = append(streamingSeriesBatches, labelsBatch)
+			}
+
+			if resp.IsEndOfSeriesStream {
+				if streamingSeriesCount > 0 {
+					result.streamingSeries.Series = make([]labels.Labels, 0, streamingSeriesCount)
+
+					for _, batch := range streamingSeriesBatches {
+						result.streamingSeries.Series = append(result.streamingSeries.Series, batch...)
+					}
+
+					streamReader := ingester_client.NewSeriesChunksStreamReader(stream, streamingSeriesCount, queryLimiter)
+					closeStream = false
```

Comment:
If an error occur on `stream.Recv()` after this point, the `queryIngester()` function will return error and the stream will **not** be closed. I think it should be closed whenever an error is returned.

Reply:
`stream.Recv()` won't be called again inside `queryIngester()` - we return unconditionally from `queryIngester()` a few lines later.